golint plugin fails silently

The golint plugin logs an error, but doesn't react on the PR if an error is encountered while identifying lint problems (like on this line: https://github.com/kubernetes/test-infra/blob/master/prow/plugins/golint/golint.go#L153).
ref: https://github.com/kubernetes/test-infra/pull/10628#issuecomment-452044803

```
linting prow/plugins/golint/suggestion/testdata/range.go: prow/plugins/golint/suggestion/testdata/range.go:22:1: expected declaration, found 'for'
```

Ideally the plugin would expose these errors to the user along with successfully identified lint problems, but at minimum the plugin should fail and post a comment with the error message.


/kind bug
/area prow/plugins
cc @taragu

--- a/prow/plugins/golint/golint.go
+++ b/prow/plugins/golint/golint.go
@@ -139,22 +139,51 @@ func newProblems(cs []github.ReviewComment, ps map[string]map[int]lint.Problem) 
 
 // problemsInFiles runs golint on the files. It returns a map from the file to
 // a map from the line in the patch to the problem.
-func problemsInFiles(r *git.Repo, files map[string]string) (map[string]map[int]lint.Problem, error) {
+func problemsInFiles(r *git.Repo, files map[string]string) (map[string]map[int]lint.Problem, []github.DraftReviewComment) {
 	problems := make(map[string]map[int]lint.Problem)
+	var lintErrorComments []github.DraftReviewComment
 	l := new(lint.Linter)
 	for f, patch := range files {
 		problems[f] = make(map[int]lint.Problem)
 		src, err := ioutil.ReadFile(filepath.Join(r.Dir, f))
 		if err != nil {
-			return nil, err
+			lintErrorComments = append(lintErrorComments, github.DraftReviewComment{
+				Path: f,
+				Body: fmt.Sprintf("%v", err),
+			})
 		}
 		ps, err := l.Lint(f, src)
 		if err != nil {
-			return nil, fmt.Errorf("linting %s: %v", f, err)
+			// Get error line by parsing the error message
+			errLineIndexStart := strings.LastIndex(err.Error(), f) + len(f)
+			reNumber := regexp.MustCompile(`:([0-9]+):`)
+			matches := reNumber.FindStringSubmatch(err.Error()[errLineIndexStart:])
+			newComment := github.DraftReviewComment{
+				Path: f,
+				Body: err.Error(),
+			}
+			if len(matches) > 1 {
+				errLineString := matches[1]
+				errLine, errAtoi := strconv.Atoi(errLineString)
+				if errAtoi == nil {
+					newComment.Position = errLine
+				}
+				// Trim error message to after the line and column numbers
+				reTrimError := regexp.MustCompile(`(:[0-9]+:[0-9]+: )`)
+				matches = reTrimError.FindStringSubmatch(err.Error())
+				if len(matches) > 0 {
+					newComment.Body = err.Error()[len(matches[0])+errLineIndexStart:]
+				}
+			}
+			lintErrorComments = append(lintErrorComments, newComment)
 		}
 		al, err := AddedLines(patch)
 		if err != nil {
-			return nil, fmt.Errorf("computing added lines in %s: %v", f, err)
+			lintErrorComments = append(lintErrorComments,
+				github.DraftReviewComment{
+					Path: f,
+					Body: fmt.Sprintf("computing added lines in %s: %v", f, err),
+				})
 		}
 		for _, p := range ps {
 			if pl, ok := al[p.Position.Line]; ok {
@@ -162,7 +191,7 @@ func problemsInFiles(r *git.Repo, files map[string]string) (map[string]map[int]l
 			}
 		}
 	}
-	return problems, nil
+	return problems, lintErrorComments
 }
 
 func handle(minimumConfidence float64, ghc githubClient, gc *git.Client, log *logrus.Entry, e *github.GenericCommentEvent) error {
@@ -210,7 +239,7 @@ func handle(minimumConfidence float64, ghc githubClient, gc *git.Client, log *lo
 	log.WithField("duration", time.Since(startClone)).Info("Cloned and checked out PR.")
 
 	// Compute lint errors.
-	problems, err := problemsInFiles(r, modifiedFiles)
+	problems, lintErrorComments := problemsInFiles(r, modifiedFiles)
 	if err != nil {
 		return err
 	}
@@ -224,14 +253,17 @@ func handle(minimumConfidence float64, ghc githubClient, gc *git.Client, log *lo
 	}
 	log.WithField("duration", time.Since(finishClone)).Info("Linted.")
 
-	oldComments, err := ghc.ListPullRequestComments(org, repo, e.Number)
-	if err != nil {
-		return err
+	nps := problems
+	if len(problems) > 0 {
+		oldComments, err := ghc.ListPullRequestComments(org, repo, e.Number)
+		if err != nil {
+			return err
+		}
+		nps = newProblems(oldComments, problems)
 	}
-	nps := newProblems(oldComments, problems)
 
 	// Make the list of comments.
-	var comments []github.DraftReviewComment
+	var comments []github.DraftReviewComment = lintErrorComments
 	for f, ls := range nps {
 		for l, p := range ls {
 			var suggestion = suggestion.SuggestCodeChange(p)

--- a/prow/plugins/golint/golint_test.go
+++ b/prow/plugins/golint/golint_test.go
@@ -425,6 +425,158 @@ func TestLintCodeSuggestion(t *testing.T) {
 	}
 }
 
+func TestLintError(t *testing.T) {
+	var testcases = []struct {
+		name       string
+		codeChange string
+		pullFiles  map[string][]byte
+		comments   []github.DraftReviewComment
+	}{
+		{
+			name: "Golint error",
+			codeChange: "@@ -0,0 +1,15 @@\n+" +
+				"/*\n+" +
+				"Package bar comment\n+" +
+				"*/\n+" +
+				"package bar\n+" +
+				"\n+" +
+				"func f(m []string) {\n+" +
+				"      for _ = range m {\n+" +
+				"      }\n+" +
+				"}\n+" +
+				"\n+" +
+				"for b() error {\n+" +
+				"return nil\n+" +
+				"}\n+",
+			pullFiles: map[string][]byte{
+				"qux.go": []byte("/*\n" +
+					"Package bar comment\n" +
+					"*/\n" +
+					"package bar\n" +
+					"\n" +
+					"func f(m []string) {\n" +
+					"      for _ = range m {\n" +
+					"      }\n" +
+					"}\n" +
+					"\n" +
+					"for b() error {\n" +
+					"return nil\n" +
+					"}\n"),
+			},
+			comments: []github.DraftReviewComment{
+				{
+					Body:     "expected declaration, found 'for'",
+					Position: 11,
+					Path:     "qux.go",
+				},
+				{
+					Body: "computing added lines in qux.go: invalid patch: " +
+						"@@ -0,0 +1,15 @@\n+" +
+						"/*\n+" +
+						"Package bar comment\n+" +
+						"*/\n+" +
+						"package bar\n+" +
+						"\n+" +
+						"func f(m []string) {\n+" +
+						"      for _ = range m {\n+" +
+						"      }\n+" +
+						"}\n+" +
+						"\n+" +
+						"for b() error {\n+" +
+						"return nil\n+" +
+						"}\n+",
+					Position: 0,
+					Path:     "qux.go",
+				},
+			},
+		},
+		{
+			name: "No golint error",
+			codeChange: "@@ -0,0 +1,9 @@\n+" +
+				"/*\n+" +
+				"Package bar comment\n+" +
+				"*/\n+" +
+				"package bar\n+" +
+				"\n+" +
+				"func f() {\n+" +
+				"      var myZeroFlt float32 = 0.\n+" +
+				"}\n+",
+			pullFiles: map[string][]byte{
+				"qux.go": []byte("/*\n" +
+					"Package bar comment\n" +
+					"*/\n" +
+					"package bar\n" +
+					"\n" +
+					"func b() error {\n" +
+					"return nil\n" +
+					"}\n"),
+			},
+			comments: nil,
+		},
+	}
+
+	lg, c, err := localgit.New()
+	if err != nil {
+		t.Fatalf("Making localgit: %v", err)
+	}
+	defer func() {
+		if err := lg.Clean(); err != nil {
+			t.Errorf("Cleaning up localgit: %v", err)
+		}
+		if err := c.Clean(); err != nil {
+			t.Errorf("Cleaning up client: %v", err)
+		}
+	}()
+	if err := lg.MakeFakeRepo("foo", "bar"); err != nil {
+		t.Fatalf("Making fake repo: %v", err)
+	}
+	if err := lg.AddCommit("foo", "bar", initialFiles); err != nil {
+		t.Fatalf("Adding initial commit: %v", err)
+	}
+	if err := lg.CheckoutNewBranch("foo", "bar", "pull/42/head"); err != nil {
+		t.Fatalf("Checking out pull branch: %v", err)
+	}
+
+	for _, test := range testcases {
+		t.Logf("Running test case %q...", test.name)
+		if err := lg.AddCommit("foo", "bar", test.pullFiles); err != nil {
+			t.Fatalf("Adding PR commit: %v", err)
+		}
+		gh := &ghc{
+			changes: []github.PullRequestChange{
+				{
+					Filename: "qux.go",
+					Patch:    test.codeChange,
+				},
+			},
+		}
+		if err := handle(0, gh, c, logrus.NewEntry(logrus.New()), e); err != nil {
+			t.Fatalf("Got error from handle: %v", err)
+		}
+
+		if test.comments == nil {
+			if len(gh.comment.Comments) > 0 {
+				t.Fatalf("Expected no comment, got %d: %v.", len(gh.comment.Comments), gh.comment.Comments)
+			}
+		} else {
+			if len(gh.comment.Comments) != len(test.comments) {
+				t.Fatalf("Expected one comments, got %d: %v.", len(gh.comment.Comments), gh.comment.Comments)
+			}
+			for _, testComment := range test.comments {
+				exists := false
+				for _, actualComment := range gh.comment.Comments {
+					if actualComment == testComment {
+						exists = true
+					}
+				}
+				if !exists {
+					t.Fatalf("Did not get back expected comment %v, actual comments are %v", testComment, gh.comment.Comments)
+				}
+			}
+		}
+	}
+}
+
 func TestAddedLines(t *testing.T) {
 	var testcases = []struct {
 		patch string

